lang/syn/src/idl/mod.rs

```diff
@@ -128,69 +129,92 @@ pub enum IdlType {
     Option(Box<IdlType>),
     Vec(Box<IdlType>),
     Array(Box<IdlType>, usize),
+    Map(Box<IdlType>, Box<IdlType>),
 }
 
 impl std::str::FromStr for IdlType {
     type Err = anyhow::Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let mut s = s.to_string();
-        fn array_from_str(inner: &str) -> IdlType {
-            match inner.strip_suffix(']') {
-                None => {
-                    let (raw_type, raw_length) = inner.rsplit_once(';').unwrap();
-                    let ty = IdlType::from_str(raw_type).unwrap();
-                    let len = raw_length.replace("_", "").parse::<usize>().unwrap();
-                    IdlType::Array(Box::new(ty), len)
+        fn syn_type_to_idl_type(syn_type: &Type) -> Result<IdlType, anyhow::Error> {
+            match syn_type {
+                Type::Path(type_path) => {
+                    let type_word = type_path.path.segments[0].ident.to_string();
+                    let type_args: Vec<Type> = {
+                        match &type_path.path.segments[0].arguments {
+                            PathArguments::AngleBracketed(x) => x
+                                .args
+                                .iter()
+                                .map(|arg| match arg {
+                                    GenericArgument::Type(ty) => Ok(ty.clone()),
+                                    _ => Err(()),
+                                })
+                                .collect::<Result<Vec<Type>, _>>(),
+                            _ => Ok(Vec::new()),
+                        }
+                        .expect("Invalid option")
+                    };
+                    let r = match type_word.as_str() {
+                        "bool" => IdlType::Bool,
+                        "u8" => IdlType::U8,
+                        "i8" => IdlType::I8,
+                        "u16" => IdlType::U16,
+                        "i16" => IdlType::I16,
+                        "u32" => IdlType::U32,
+                        "i32" => IdlType::I32,
+                        "u64" => IdlType::U64,
+                        "i64" => IdlType::I64,
+                        "u128" => IdlType::U128,
+                        "i128" => IdlType::I128,
+                        "Vec<u8>" => IdlType::Bytes,
+                        "String" => IdlType::String,
+                        "Pubkey" => IdlType::PublicKey,
+                        "Option" => {
+                            if type_args.len() != 1 {
+                                return Err(anyhow::anyhow!("Invalid option"));
+                            }
+                            let inner_ty = syn_type_to_idl_type(&type_args[0])?;
+                            IdlType::Option(Box::new(inner_ty))
+                        }
+                        "Vec" | "VecDeque" | "LinkedList" => {
```

Comment:
I also added support to `VecDeque` and `LinkedList` since they are serialized by borsh in the same way as `Vec`.